Add ormar support

collerek opened the pull request:
Hi,

Since I am lazy and want the tools to do boring work for me I installed and used:

*  `flake8` - to check styles, pep8, complexity etc. with set of plugins to check different style related issues
*  `black`  - installed with `flake8` (one of plugins) to format code in a consistent way -> always the same regardless who is making the change, I especially like how it breaks long lines
*  `mypy` - to check the types annotations

I also modified the code to pass all those test/ tools which mean a lot of minor changes basically everywhere, added missing types, fixed types enough to pass `mypy` checks (although it's not done fully), remove unused imports, reorder import etc. You still have few minor issues like some too complex methods according to those checks but that I didn't want to change.

I also extracted creation of the items in pagination tests into `fixtures` and changed it's scope (scope is now class so they run once per implementation), as before they were running per every parametrization and in effect the tests for all backends were running **several minutes** and now they run like 40s, and tests should run quickly or you end up not running them often enough. Since you test creation in the separate tests I think it's beneficial as you do not create the same items 20-30 times in those few tests.

I know that you might not like such a massive changes (although **they are all style related - there is no logic change** - apart from `ormar` implementation of course :) ), so if you refuse it we can just extract `ormar` files (tests and backend) with docs. 

It was still easier for me to check my `ormar` implementation this way even if you would want to discard those changes. 

But I think the project would benefit from tiding up that was done in this PR, so please review the PR and let me know what you think.

awtkns commented on tests/conftest.py:
```diff
@@ -4,11 +4,11 @@
 
 
-@pytest.fixture(params=implementations)
+@pytest.fixture(params=implementations, scope="session")
```
```suggestion
@pytest.fixture(params=implementations, scope="function")
```
There should be a clean testing environment per test. Default is function.

collerek replied:
In general I agree, but later you use the same client to insert data and since in pagination test i refactored it to be `class` fixture to not insert it multiple times as you only fetch data, and fetch from the same data, in your test, the `class` scope fixture cannot depend on `function` scope one. 

Pagination fixtures can be changed to `function` scope too, but then it will be much slower, it's still much faster then before as before you run the inserts not per function but per parametrization. 

Alternative is to insert data in a different way for pagination purposes, although if client will clean the data on each run (and is function scope) it's still have to be function scope. (I'm not sure if you clean the data in client?)

awtkns replied:
I agree. I have changed it to scope class which seems to be the best of both worlds.

awtkns commented:
Thanks for setting up Flake and Black. I've been meaning to get around to that. I do agree that having a consistent code style would be of great benefit to the project!  Hoping to merge soon, I just want to take a closer look later tonight.

awtkns commented:
LGTM! Thanks so much for all the great work. The Ormar backend will be release once this is merged. I just have to write some release notes first.

By the way, what is the maturity of this project, will there be breaking changes in the future?

awtkns commented:
Actually @collerek I have one more question. Correct me if I am wrong, but isn't one of the main selling points of ormar is the fact that you don't have to maintain both a pydantic and database model.  However, in the OrmarCRUDRouter we are requiring the user to pass both a db_model and a schema. It might be better if we removed the db_model kwarg and passed the Ormar model as the 'schema'. Or vice versa. 

Thoughts?

collerek commented:
Haha,, good point, that's what you end up if you copy too much from other backends 😂 for autoincrement or pk with default function you do not require the create model as with ormar they are optional and pydantic will not complain. So it should work pretty much straight out of the box. You prefer to keep schema or db model?

collerek commented:
OK, so you can use just `ormar` but then the autoincrement primary key fields become optional, yet they are still there (as you very can overwrite autoincrement pk if needed). And since you actually pass the `id=0` in your sample (which I also don't like as usually it should start from 1 not 0, but I assume you had to pass any integer in order to create `pydantic` model as field was required).

As in I think 100% of test you later call `model.dict()` to be able to either send the data or compare it to response I simply changed them to `dict`s (and adjusted tests). That way the useless `id` can be dropped, and id is not explicitly passed to API (unless required like in string pk tests. I think it's also cleaner as you don't have to explicitly call `dict()` so many times)

I unified the name of the primary key field in backends, as it was different in some of them (now it's always `self._pk`, maybe it should also be set explicitly in mem backend), and the actual name of the pk was not passed to schema generation function, when the custom id was used (it was always hardcoded `'id'`).

Let me know what you think about the changes.

awtkns commented:
It all looks good to me on my end. Thank so much for your help!

collerek commented:
Happy to help and keep up the great work! 
Let me know if you want me to do anything else before you can merge this :)

awtkns commented:
My last concern before we merge this is the potential for database IntegrityError's in the create and update routes.  In the sqlalchemy router as an example, we try catch both the create and update queries (Returning a 422 should it fail). Not sure how this is handled in Ormar.

**fastapi_crudrouter/core/sqlalchemy.py**
```python
    def _create(self, *args: Any, **kwargs: Any) -> Callable:
        def route(
            model: self.create_schema,  # type: ignore
            db: Session = Depends(self.db_func),
        ) -> TM:
            try:
                db_model = self.db_model(**model.dict())
                db.add(db_model)
                db.commit()
                db.refresh(db_model)
                return db_model
            except IntegrityError:
                db.rollback()
                raise HTTPException(422, "Key already exists")

        return route

    def _update(self, *args: Any, **kwargs: Any) -> Callable:
        def route(
            item_id: self._pk_type,  # type: ignore
            model: self.update_schema,  # type: ignore
            db: Session = Depends(self.db_func),
        ) -> TM:
            try:
                db_model = self._get_one()(item_id, db)

                for key, value in model.dict(exclude={self._pk}).items():
                    if hasattr(db_model, key):
                        setattr(db_model, key, value)

                db.commit()
                db.refresh(db_model)

                return db_model
            except IntegrityError as e:
                db.rollback()
                raise HTTPException(422, ", ".join(e.args))

        return route
```

Thoughts? After this I am good to merge.

collerek commented:
Now it does nothing :) 

So will just throw IntegrityError (db backend specific, so one of different 3 for sqlite, mysql and pg). 
You want to return 422 in this case?

awtkns commented:
> Now it does nothing :)
> 
> So will just throw IntegrityError (db backend specific, so one of different 3 for sqlite, mysql and pg).
> You want to return 422 in this case?

Yup that sounds good to me. #49

awtkns commented on docs/en/docs/backends/ormar.md:
```diff
@@ -0,0 +1,86 @@
+When generating routes, the `OrmarCRUDRouter` will automatically tie into your database
+using your [ormar](https://collerek.github.io/ormar/) models. To use it, you must pass a
+[pydantic](https://pydantic-docs.helpmanual.io/) schema and your `ormar` database model to
+it.
```
```suggestion
using your [ormar](https://collerek.github.io/ormar/) models. To use it, simply pass your ormar database model as the schema.
```

collerek commented:
Ok I: 

*  added IntegrityError to ormar for all database backends.
* checked ormar backend also with mysql and postgress -> that's why I changed TestClient to context_managers, that way startup/shutdown events are triggered and i.e. db is connected (this is not required for sqlite but is for other db backends).
*  Refactored sqlalchemy integrity test into a common one, as you probably would want to add this for other backends too, and now it's a new implementation/client as other tests.

awtkns commented:
Awesome thanks for all the help. Great work as always.  Been meaning to refactor out the integrity error tests so thanks for that
